feat: Transfer logs from a process server to a central system

DeboraTahara opened the pull request.

eduardacoppo committed:
chore: move log transfer to log_runtime_archive

eduardacoppo committed:
test: add tests for transfer and watch_transfer

eduardacoppo committed:
chore: general fixes and variable renaming

eduardacoppo committed:
chore: fix rubocop offenses

DeboraTahara committed:
fix: Moving the FTPUpload to syskit/runtime/remote/server diretory

DeboraTahara committed:
fix: refactoring the process_dataset_transfer to use the FTPUpload

doudou commented on lib/syskit/cli/log_runtime_archive.rb:
```diff
@@ -29,6 +31,19 @@ def initialize(
+            # Iterate over all datasets in a Roby log root folder and transfer them
+            # through FTP server
+            #
+            # @param [Params] server_params the FTP server parameters:
```
use a Struct instead of a hash

`FTPParameters = Struct.new(:host, :port, ..., keyword_init: true)`

doudou commented on lib/syskit/cli/log_runtime_archive.rb:
```diff
@@ -29,6 +31,19 @@ def initialize(
+                candidates = self.class.find_all_dataset_folders(@root_dir)
+                candidates.each do |child|
+                    basename = child.basename.to_s
+                    process_dataset_transfer(basename, server_params)
```
`basename` here points to the folder basename, not a file name (`candidates` is a `Array<Pathname>` that points to folders as the name "find_all_dataset_folders" indicates). Beyond needing the full path and not the base name, FTPUpload works at the file level.

However, the whole complexity of this class is that we deal with folders where some files can be copied (the ones that are "finished" and some not (the ones that are being updated). The whole logic of what can be processed or not is currently in `archive_dataset`. You need to refactor it to share that logic here in the transfer case, but transferring instead of archiving.

doudou commented on lib/syskit/runtime/remote/server/ftp_upload.rb:
```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
```
What is this folder ?

I would put it in `roby_app/log_transfer` to be honest. And move SpawnServer here (renaming it into just Server)

DeboraTahara committed:
chore: Reorganize files

DeboraTahara committed:
fix: Create FTPParameters struct

DeboraTahara committed:
fix: Change the spawn_server path required

doudou commented on lib/syskit/cli/log_runtime_archive.rb:
```diff
@@ -2,6 +2,9 @@
 
 require "archive/tar/minitar"
 require "sys/filesystem"
+require "syskit/process_managers/remote/protocol"
```
You shouldn't need that one ...

doudou commented on lib/syskit/cli/log_runtime_archive.rb:
```diff
@@ -2,6 +2,9 @@
 require "sys/filesystem"
+require "syskit/process_managers/remote/protocol"
+require "syskit/roby_app/log_transfer_server/ftp_upload"
+require "net/ftp"
```
And you should definitely not require this one explicitly. If it is needed, it should be required by `ftp_upload`.

DeboraTahara committed:
fix: Refactor the process_root_folder_transfer

doudou commented on lib/syskit/cli/log_runtime_archive.rb:
```diff
@@ -104,6 +123,46 @@ def process_dataset(child, full:)
+                logger.info(
+                    "Transfering dataset #{dataset_path} in #{full ? 'full' : 'partial'} mode"
+                )
+                candidates = each_file_from_path(dataset_path).to_a
```
Not needed anymore I believe ...

DeboraTahara replied:
Why not? We still need to have a list of all files.

doudou commented on test/cli/test_log_runtime_archive.rb:
```diff
@@ -516,14 +516,73 @@ def should_archive_dataset(dataset, archive_basename, full:)
+                end
+
+                def transfer_dataset(ftp, src_path, tgt_path)
+                    ftp.putbinaryfile(src_path, tgt_path)
```
I'm pretty sure that's not what you want to do in this test ...

doudou reviewed (commented):
From where I sit, I think the general structure is valid. Missing tests, but I'm guessing you are aware of that ...

DeboraTahara committed:
fix: Reorganizing files

DeboraTahara committed:
fix: Creating unit tests for FTP

eduardacoppo committed:
chore: minor fixes and rubocop fixes

eduardacoppo committed:
fix: convert data_dir to string

Ftpd base methods expect data_dir to be a string

eduardacoppo committed:
chore: minor fixes and changes

eduardacoppo committed:
fix: creates dataset folder(s) if it does not exist

eduardacoppo committed:
chore: fix rubocop offenses

eduardacoppo committed:
fix: rewrite and fix tests, minor fixes and cleanup

eduardacoppo commented on lib/syskit/cli/log_runtime_archive.rb:
```diff
@@ -39,10 +39,10 @@ def initialize(
             #
             # @param [Params] server_params the FTP server parameters
             def process_root_folder_transfer(server_params)
-                candidates = self.class.find_all_dataset_folders(@root_dir)
```
Just noticed this should be kept. The folder names in the tests should match the pattern. Will fix it.

eduardacoppo committed:
fix: re-add function and fix tests

re-add find_all_dataset_folders function and rewrite dataset folders in the root folder transfer test to match the pattern

eduardacoppo committed:
fix: add error to rescue

eduardacoppo committed:
chore: rename test to make it clearer

jhonasiv commented on lib/syskit/roby_app/log_transfer_server/ftp_upload.rb:
```diff
@@ -0,0 +1,111 @@
+                    dataset_path = File.dirname(@file.relative_path_from(root))
+
+                    dataset_path.split("/") do |folder|
+                        ftp.chdir(folder)
+                    rescue StandardError => e
+                        ftp.mkdir(folder)
+                        ftp.chdir(folder)
```
Do not rescue this, its better to look for the folder instead, I guess using ftp.ls

jhonasiv replied:
Actually, after taking a deeper look into this, using ls would make WriteOnlyFileSystem perform a read...and it would also increase the number of ftp requests per candidate to at least 2. I now think that rescuing the chdir is better.

jhonasiv commented on lib/syskit/runtime/server/write_only_disk_file_system.rb:
```diff
@@ -1,16 +1,18 @@
 
                 def initialize(data_dir)
-                    set_data_dir data_dir
+                    # Ftpd base methods expect data_dir to be a string
```
You can raise if data_dir doesnt support #to_s, explaining this expectation

jhonasiv commented on test/cli/test_log_runtime_archive.rb:
```diff
@@ -516,14 +517,121 @@ def should_archive_dataset(dataset, archive_basename, full:)
+                        assert(File.exist?(@target_dir / "20220434-2023" / "test.0.log"))
+                        assert(File.exist?(@target_dir / "20220434-2023" / "test.1.log"))
+                        assert(File.exist?(@target_dir / "20220434-2024" / "test.0.log"))
+                        refute(File.exist?(@target_dir / "20220434-2024" / "test.1.log"))
```
Why wouldnt `test.1.log` be for dataset_b be finished in this test?

jhonasiv commented on test/cli/test_log_runtime_archive_main.rb:
```diff
@@ -128,6 +129,118 @@ def call_archive(root_path, archive_path, low_limit, freed_limit)
+                end
+
+                it "creates an FTP server" do
+                    flexmock(Runtime::Server::SpawnServer)
+                        .new_instances
+                        .should_receive(:initialize).with(
+                            @target_dir, *@server_params.values
+                        ).once
```
This doesnt test anything, I am not even sure how this passes.IMO the best way to test that the server was created is to send a noop command to it after starting it normally through the CLI.

jhonasiv replied:
Opening the connection with the server is fine too

jhonasiv commented on test/cli/test_log_runtime_archive_main.rb:
```diff
@@ -128,6 +129,118 @@ def call_archive(root_path, archive_path, low_limit, freed_limit)
+
+                it "calls transfer with the specified period" do
+                    mock_files_size([])
+                    mock_available_space(200) # 70 MB
```
Comment is wrong

eduardacoppo committed:
chore: add comment explaining why last log is not transfered

eduardacoppo committed:
chore: remove unnecessary code

eduardacoppo committed:
chore: raise ArgumentError if data_dir is not convertible

eduardacoppo committed:
chore: fix rubocop offenses and respond_to arg

eduardacoppo committed:
fix: fix transfer_server test

eduardacoppo committed:
fix: add root as optional argument

eduardacoppo committed:
chore: fix rubocop offenses

jhonasiv commented on lib/syskit/roby_app/log_transfer_server/ftp_upload.rb:
```diff
@@ -0,0 +1,112 @@
+                    LogUploadState::Result.new(@file, false, e.message)
+                end
+
+                def ensure_dataset_path_exists(ftp, root)
```
```suggestion
                def chdir_to_file_directory(ftp, root)
```

wvmcastro replied:
rt

wvmcastro commented on lib/syskit/cli/log_runtime_archive.rb:
```diff
@@ -17,8 +18,12 @@ class CompressionFailed < RuntimeError; end
+
             def initialize(
-                root_dir, target_dir,
+                root_dir, target_dir: nil,
```
Document how `target_dir` impact `LogRuntimeArchive` behaviour. It is obvious now, but it won't be once we forget about it

jhonasiv commented on lib/syskit/roby_app/log_transfer_server/ftp_upload.rb:
```diff
@@ -0,0 +1,112 @@
+                    last = Time.now
+                    opening_mode = root ? "w+" : "r"
+                    File.open(@file, opening_mode) do |file_io|
+                        ensure_dataset_path_exists(ftp, root) if root
```
```suggestion
                    ensure_dataset_path_exists(ftp, root) if root
                    File.open(@file) do |file_io|                        
```

wvmcastro commented on lib/syskit/cli/log_runtime_archive.rb:
```diff
@@ -83,9 +102,10 @@ def ensure_free_space(free_space_low_limit, free_space_delete_until)
 
             def process_dataset(child, full:)
                 use_existing = true
+                basename = child.basename.to_s
```
if you really want to create a variable, I would name it `filename` instead of `basename`. It is more descriptive in my opinion.

wvmcastro commented on lib/syskit/roby_app/log_transfer_server/ftp_upload.rb:
```diff
@@ -0,0 +1,112 @@
+                # @param [Pathname] root the archive root folder
+                def transfer(ftp, root)
+                    last = Time.now
+                    opening_mode = root ? "w+" : "r"
```
It is not clear for me why the opening mode changes conditioned by root?

wvmcastro commented on lib/syskit/roby_app/log_transfer_server/ftp_upload.rb:
```diff
@@ -0,0 +1,112 @@
+
+                    dataset_path.split("/") do |folder|
+                        ftp.chdir(folder)
+                    rescue StandardError => _e
```
This _rescue_ will catch anything, I think we want to catch just FTP related exceptions, i.e. `FTPPermError` https://ruby-doc.org/stdlib-2.4.0/libdoc/net/ftp/rdoc/Net/FTP.html#method-i-chdir

eduardacoppo committed:
chore: rescue specific error

eduardacoppo committed:
fix: specify rate in mbps and convert to bps when needed

eduardacoppo committed:
chore: ensure path exists before opening file

eduardacoppo committed:
chore: remove unnecessary variable

wvmcastro commented on lib/syskit/roby_app/log_transfer_server/ftp_upload.rb:
```diff
@@ -0,0 +1,111 @@
+                # @param [Pathname] root the archive root folder
+                def transfer(ftp, root)
+                    last = Time.now
+                    ensure_dataset_path_exists(ftp, root) if root
```
it is not clear when root can be _nil_, in fact I would expect it should never be _nil_

eduardacoppo replied:
Already answered this in private. But just for the record:
`root` is `nil` in the usage that already existed. It wasn't used before and this can be easily checked by just reading the code.
Why I added this optional argument: The FTP server only copies the file, it does not create the hierarchy of folders the file is under. And we must create this hierarchy, otherwise all log files from all datasets will be in the same place (and some would possibly be overwritten). So I pass root to know exactly the folder hierarchy the file is under, create these folders and copy the file to the correct location.

wvmcastro replied:
> It wasn't used before

Exactly the point I was trying to make. If it is not used why bother maintaining the behaviour?

jhonasiv replied:
> > It wasn't used before
> 
> Exactly the point I was trying to make. If it is not used why bother maintaining the behaviour?

Because the parameter was introduced in this PR. There is still parts of syskit that use the #transfer method so they should continue working.

wvmcastro reviewed (commented):
Did yo forget https://github.com/rock-core/tools-syskit/pull/448#discussion_r1913135056 ? Pass through all comments so far, also mark the resolved ones as resolved, please

eduardacoppo committed:
chore: change function name, document target_dir

and fix rubocop offenses

wvmcastro commented on lib/syskit/cli/log_runtime_archive.rb:
```diff
@@ -17,8 +18,19 @@ class CompressionFailed < RuntimeError; end
+
+            # Initializes the LogRuntimeArchive
+            #
+            # @param [Pathname] root_dir the file's source path
+            # @param [Pathname] target_dir the path to store the file in the archive,
+            #   which is nil in transfer mode, as the target_dir is the FTP server
```
```suggestion
            # @param [Pathname] root_dir the logs directory
            # @param [Pathname] target_dir the path to store the file in the archive,
            #   which should be nil in transfer mode, as the logs will be transferred directly to the ftp server @see process_root_folder_transfer
```
